# Data Preprocessing

✔ Missing value handling
✔ Duplicate removal
✔ Data type fixing
✔ Outlier removal
✔ Logical data validation
✔ Dataset standardization

In [3]:
import pandas as pd 

In [4]:
df=pd.read_csv("AB_NYC_2019.csv")

In [5]:
df.head(3)

id                                 name  host_id  host_name  \
0  2539   Clean & quiet apt home by the park     2787       John   
1  2595                Skylit Midtown Castle     2845   Jennifer   
2  3647  THE VILLAGE OF HARLEM....NEW YORK !     4632  Elisabeth   

  neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0            Brooklyn    Kensington  40.64749  -73.97237     Private room   
1           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt   
2           Manhattan        Harlem  40.80902  -73.94190     Private room   

   price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0    149               1                  9  2018-10-19               0.21   
1    225               1                 45  2019-05-21               0.38   
2    150               3                  0         NaN                NaN   

   calculated_host_listings_count  availability_365  
0                               6               365  
1                               2               355  
2                               1               365

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

# removing Null values

In [7]:
df.isnull().sum()

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

## WE are removing nullvalues in name,hosthame as well also replacing  last_review   10052
### reviews_per_month  10052 to 0 

In [10]:


df = df.dropna(subset=['name', 'host_name'])

# Fill missing values by -0
df['reviews_per_month'] = df['reviews_per_month'].fillna(0)
df['last_review'] = df['last_review'].fillna(0)


In [11]:
df.isnull().sum()

id                                0
name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
dtype: int64

In [12]:
# now dublicates

In [13]:
df.drop_duplicates(inplace=True)

df.duplicated().sum()


np.int64(0)

In [15]:
df['last_review'] = pd.to_datetime(df['last_review'], errors='coerce')

In [16]:
# data types 

In [17]:
df.dtypes


id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                      int64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
dtype: object

# statisticaal info

In [19]:
df['price'].describe()


count    48858.000000
mean       152.740309
std        240.232386
min          0.000000
25%         69.000000
50%        106.000000
75%        175.000000
max      10000.000000
Name: price, dtype: float64

In [22]:
# logical errors
df = df[df['price'] > 0]
df = df[df['minimum_nights'] > 0]


# str.strip() removes any extra spaces from the start and end of text in each cell.

In [23]:
df['neighbourhood_group'] = df['neighbourhood_group'].str.strip()
df['room_type'] = df['room_type'].str.strip()


In [24]:
df['neighbourhood_group'].value_counts()


neighbourhood_group
Manhattan        21642
Brooklyn         20080
Queens            5664
Bronx             1088
Staten Island      373
Name: count, dtype: int64

### now resewting index as we have removed sum nll instances

In [26]:
df.reset_index(drop=True, inplace=True)
df

id                                               name   host_id  \
0          2539                 Clean & quiet apt home by the park      2787   
1          2595                              Skylit Midtown Castle      2845   
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
3          3831                    Cozy Entire Floor of Brownstone      4869   
4          5022   Entire Apt: Spacious Studio/Loft by central park      7192   
...         ...                                                ...       ...   
48842  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48843  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48844  36485431            Sunny Studio at Historical Neighborhood  23492952   
48845  36485609               43rd St. Time Square-cozy single bed  30985759   
48846  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

           host_name neighbourhood_group       neighbourhood  latitude  \
0               John            Brooklyn          Kensington  40.64749   
1           Jennifer           Manhattan             Midtown  40.75362   
2          Elisabeth           Manhattan              Harlem  40.80902   
3        LisaRoxanne            Brooklyn        Clinton Hill  40.68514   
4              Laura           Manhattan         East Harlem  40.79851   
...              ...                 ...                 ...       ...   
48842        Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48843        Marisol            Brooklyn            Bushwick  40.70184   
48844  Ilgar & Aysel           Manhattan              Harlem  40.81475   
48845            Taz           Manhattan      Hell's Kitchen  40.75751   
48846     Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0      -73.97237     Private room    149               1                  9   
1      -73.98377  Entire home/apt    225               1                 45   
2      -73.94190     Private room    150               3                  0   
3      -73.95976  Entire home/apt     89               1                270   
4      -73.94399  Entire home/apt     80              10                  9   
...          ...              ...    ...             ...                ...   
48842  -73.94995     Private room     70               2                  0   
48843  -73.93317     Private room     40               4                  0   
48844  -73.94867  Entire home/apt    115              10                  0   
48845  -73.99112      Shared room     55               1                  0   
48846  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2018-10-19               0.21                               6   
1      2019-05-21               0.38                               2   
2             NaT               0.00                               1   
3      2019-07-05               4.64                               1   
4      2018-11-19               0.10                               1   
...           ...                ...                             ...   
48842         NaT               0.00                               2   
48843         NaT               0.00                               2   
48844         NaT               0.00                               1   
48845         NaT               0.00                               6   
48846         NaT               0.00                               1   

       availability_365  
0                   365  
1                   355  
2                   365  
3                   194  
4                     0  
...                 ...  
48842                 9  
48843                36  
48844                27  
48845                 2  
48846                23  

[48847 rows x 16 columns]

In [27]:
df.isnull().sum()
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48847 entries, 0 to 48846
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              48847 non-null  int64         
 1   name                            48847 non-null  object        
 2   host_id                         48847 non-null  int64         
 3   host_name                       48847 non-null  object        
 4   neighbourhood_group             48847 non-null  object        
 5   neighbourhood                   48847 non-null  object        
 6   latitude                        48847 non-null  float64       
 7   longitude                       48847 non-null  float64       
 8   room_type                       48847 non-null  object        
 9   price                           48847 non-null  int64         
 10  minimum_nights                  48847 non-null  int64         
 11  nu

In [30]:
#df.to_csv("airbnb_cleaned.csv", index=False)


In [31]:
 df.head(3)

id                                 name  host_id  host_name  \
0  2539   Clean & quiet apt home by the park     2787       John   
1  2595                Skylit Midtown Castle     2845   Jennifer   
2  3647  THE VILLAGE OF HARLEM....NEW YORK !     4632  Elisabeth   

  neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0            Brooklyn    Kensington  40.64749  -73.97237     Private room   
1           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt   
2           Manhattan        Harlem  40.80902  -73.94190     Private room   

   price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0    149               1                  9  2018-10-19               0.21   
1    225               1                 45  2019-05-21               0.38   
2    150               3                  0         NaT               0.00   

   calculated_host_listings_count  availability_365  
0                               6               365  
1                               2               355  
2                               1               365